File: Artificial_Neural_Network/Artificial_Neural_Network.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jul 23 21:34:18 2019

@author: Venkatesh Ravichandran

Artificial Neural Network

"""
#import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#import dataset
df = pd.read_csv('Churn_Modelling.csv')

#assign dependent and independent variables
X = df.iloc[:, 3:13].values
y = df.iloc[:,13].values

#Treating categorical variables
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelEncoder1 = LabelEncoder()
X[:,1] = labelEncoder1.fit_transform(X[:,1])
labelEncoder2 = LabelEncoder()
X[:,2] = labelEncoder2.fit_transform(X[:,2])

onehotEncoder = OneHotEncoder(categorical_features = [1])
X = onehotEncoder.fit_transform(X).toarray()
X = X[:,1:]

#splitting into train and test set
from sklearn.model_selection import train_test_split
train_x, test_x, train_y, test_y = train_test_split(X, y, test_size = 0.2, random_state = 0)

#Feature scaling
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
train_x = sc_x.fit_transform(train_x)
test_x = sc_x.transform(test_x)

# Building Artificial Neural Network(ANN model)
import keras
from keras.models import Sequential
from keras.layers import Dense

#Building the input and first hidden layer
classifier = Sequential()
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 11))

#Adding the second hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu'))

#Adding the output layer
classifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))

classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

classifier.fit(train_x, train_y,batch_size = 10, nb_epoch = 100)

y_pred = classifier.predict(test_x)
y_pred = y_pred>0.5

#Confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(test_y, y_pred)

File: Artificial_Neural_Network/Artificial_Neural_Network.R
#Artificial Neural Networks

#read csv data
df = read.csv("Churn_Modelling.csv")
df = df[4:14]

#mapping categorical variables
df$Geography = as.numeric(factor(df$Geography,
                                      levels = c('France', 'Spain', 'Germany'),
                                      labels = c(1, 2, 3)))
df$Gender = as.numeric(factor(df$Gender,
                              levels = c('Male','Female'),
                              labels = c(1, 2)))

#splitting into train and test sets
library(caTools)
set.seed(100)
split = sample.split(df$Exited, SplitRatio = 0.8)
train = subset(df, split == TRUE)
test = subset(df, split == FALSE)

#Feature scaling
train[-11] = scale(train[-11])
test[-11] = scale(test[-11])

#Building ANN
#install.packages('h2o')
library(h2o)
h2o.init(nthreads = -1)

classifier = h2o.deeplearning(y = 'Exited',
                              training_frame = as.h2o(train),
                              activation = 'Rectifier',
                              hidden = c(6,6),
                              epochs = 100,
                              train_samples_per_iteration = -2)

#prediction on new data
y_pred = h2o.predict(classifier, newdata = as.h2o(test[-11]))

y_pred = (y_pred>0.5)

y_pred = as.vector(y_pred)

#Confusion matrix
cm = table(test[,11], y_pred)

#shutdown the h2o
h2o.shutdown()